photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala

```diff
@@ -14,60 +14,71 @@
  */
 package com.linkedin.photon.ml.algorithm
 
+import scala.collection.mutable
+
 import org.apache.spark.SparkContext
+import org.apache.spark.ml.linalg.{Vector => SparkVector}
 import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.{DataFrame, functions}
 import org.apache.spark.storage.StorageLevel
 
+import com.linkedin.photon.ml.Constants
+import com.linkedin.photon.ml.Types.{FeatureShardId, REType}
+import com.linkedin.photon.ml.constants.DataConst
 import com.linkedin.photon.ml.data._
 import com.linkedin.photon.ml.data.scoring.CoordinateDataScores
 import com.linkedin.photon.ml.function.SingleNodeObjectiveFunction
 import com.linkedin.photon.ml.model.{Coefficients, DatumScoringModel, RandomEffectModel}
 import com.linkedin.photon.ml.normalization.NormalizationContext
-import com.linkedin.photon.ml.optimization.game.{RandomEffectOptimizationConfiguration, RandomEffectOptimizationProblem}
-import com.linkedin.photon.ml.optimization._
 import com.linkedin.photon.ml.optimization.VarianceComputationType.VarianceComputationType
+import com.linkedin.photon.ml.optimization._
+import com.linkedin.photon.ml.optimization.game.{RandomEffectOptimizationConfiguration, RandomEffectOptimizationProblem}
 import com.linkedin.photon.ml.spark.RDDLike
 import com.linkedin.photon.ml.supervised.model.GeneralizedLinearModel
-
+import com.linkedin.photon.ml.util.VectorUtils
 /**
  * The optimization problem coordinate for a random effect model.
  *
  * @tparam Objective The type of objective function used to solve individual random effect optimization problems
- * @param dataset The training dataset
+ * @param rEType  The random effect type
+ * @param rawData    The raw training dataframe
  * @param optimizationProblem The random effect optimization problem
+ * @param inputColumnsNames
  */
 protected[ml] class RandomEffectCoordinate[Objective <: SingleNodeObjectiveFunction](
-    override protected val  dataset: RandomEffectDataset,
-    protected val optimizationProblem: RandomEffectOptimizationProblem[Objective])
-  extends Coordinate[RandomEffectDataset](dataset)
-    with ModelProjection
+  rEType: REType,
+  rawData: DataFrame,
+  optimizationProblem: RandomEffectOptimizationProblem[Objective],
+  featureShardId: FeatureShardId,
+  inputColumnsNames: InputColumnsNames)
+  extends Coordinate
     with RDDLike {
 
+  /* Get the training data from raw data */
+  var dataset: DataFrame = {
+    val label =  inputColumnsNames(InputColumnsNames.RESPONSE)
+    val offset = inputColumnsNames(InputColumnsNames.OFFSET)
+    val weight = inputColumnsNames(InputColumnsNames.WEIGHT)
+
+    rawData
+      .select(rEType, featureShardId, label, offset, weight)
+      .groupBy(rEType)
+      .agg(
+        functions.collect_list(featureShardId),
+        functions.collect_list(label),
+        functions.collect_list(offset),
+        functions.collect_list(weight))
+  }
```

Comment:
See comments in `FixedEffectCoordinate` - this is not how the design doc describes training, it should be more like the experiment code.